docs/source/user/stencil.rst

```diff
@@ -0,0 +1,205 @@
+kernel is applied.  
+Indexing into this array takes place with respect to the current element
+of the output array being processed.  For example, if element ``(x,y)``
+is being processed then ``a[0,0]`` in the stencil kernel corresponds to 
+``a[x+0,y+0]`` in the input array.  Similarly, ``a[-1,1]`` in the stencil
+kernel corresponds to ``a[x-1,y+1]`` in the input array.
+
+Depending on the specified kernel, the kernel cannot be applied to the
+borders of the output image as this may cause the input array to be
+accessed out-of-bounds.  How the stencil decorator handles this situation
+is dependent upon which :ref:`stencil-mode` is selected.  
+The default mode is for the stencil decorator to set the border elements 
+of the output array to zero.
+
+To invoke a stencil on an input array, call the stencil as if it were
+a regular function and pass the input array as the argument::
+
+   >>> import numpy as np
+   >>> input_arr = np.arange(100).reshape((10, 10))
+   array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
+          [10, 11, 12, 13, 14, 15, 16, 17, 18, 19],
+          [20, 21, 22, 23, 24, 25, 26, 27, 28, 29],
+          [30, 31, 32, 33, 34, 35, 36, 37, 38, 39],
+          [40, 41, 42, 43, 44, 45, 46, 47, 48, 49],
+          [50, 51, 52, 53, 54, 55, 56, 57, 58, 59],
+          [60, 61, 62, 63, 64, 65, 66, 67, 68, 69],
+          [70, 71, 72, 73, 74, 75, 76, 77, 78, 79],
+          [80, 81, 82, 83, 84, 85, 86, 87, 88, 89],
+          [90, 91, 92, 93, 94, 95, 96, 97, 98, 99]])
+   >>> output_arr = kernel1(input_arr)
+   array([[  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.],
+          [  0.,  11.,  12.,  13.,  14.,  15.,  16.,  17.,  18.,   0.],
+          [  0.,  21.,  22.,  23.,  24.,  25.,  26.,  27.,  28.,   0.],
+          [  0.,  31.,  32.,  33.,  34.,  35.,  36.,  37.,  38.,   0.],
+          [  0.,  41.,  42.,  43.,  44.,  45.,  46.,  47.,  48.,   0.],
+          [  0.,  51.,  52.,  53.,  54.,  55.,  56.,  57.,  58.,   0.],
+          [  0.,  61.,  62.,  63.,  64.,  65.,  66.,  67.,  68.,   0.],
+          [  0.,  71.,  72.,  73.,  74.,  75.,  76.,  77.,  78.,   0.],
+          [  0.,  81.,  82.,  83.,  84.,  85.,  86.,  87.,  88.,   0.],
+          [  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.]])
+   >>> input_arr.dtype
+   dtype('int64')
+   >>> output_arr.dtype
+   dtype('float64')
+
+Note that the stencil decorator has determined that the output type
+of the specified stencil kernel is ``float`` and has thus created the
+output array as ``float`` while the input array is of type ``int``.
+
+Stencil Parameters
+==================
+
+Stencil kernel definitions may take any number of arguments with
+the following provisions.  The first parameter must be an array.
+The size and shape of the output array will be the same as the
+first parameter.  Additional parameters may either be scalars or
+arrays.  For array parameters, those arrays must be at least as large
```

Comment:
I believe all your examples should generate errors.  We'll work on it.

Reply:
Thanks very much. Though I have to add that being able to pass in an array of weights would be a really useful behaviour, I'd have thought `k3` ought might have worked with this?

Reply:
At this point, we don't have a way to differentiate arrays that we want to use relative indexing on and those that we want to use standard indexing on.  Perhaps another parameter to @stencil that lets us specify which parameters should use standard indexing?  Suggestions for name of that parameter?

Reply:
Ahead of naming the parameter, I wonder what would it look like? Would it be a container of strings to refer to arguments, or perhaps the contents is the arguments themselves? e.g. 
```python
@stencil(relative_indexing=(a, b,), standard_indexing=(w,))
def k(a, b, w):
    return a[0, 1] + b[0, -1] + w[4, 17]
```
I guess (`relidx`/`stdidx`), (`idxrel`/`idxstd`) or similar could be short forms name wise? I think this should be open to suggestions as it's quite important.

I guess another way would be to specify in the decorator the number of args that should be assumed to use relative indexing from position zero, the rest are assumed standard.
```python
@stencil(nargs_rel_idx=2)
def k(a, b, w):
    return a[0, 1] + b[0, -1] + w[4, 17]
```
Anyone else have a view?

Reply:
Yes, the first way is what I was thinking about.  I'm not a big believer in saving characters so I would go for the long form only.

Reply:
Now that I have thought about it more, I don't like the relative_indexing option because that just creates more problems.  What if there is an array that isn't in either list?  Seems like we can just define away these problems by only have the standard_indexing option.  Everything not specified there uses relative indexing

Reply:
The standard indexing option is implemented but requires the string name of the variable like standard_indexing=("b",).

Reply:
Thanks for this fix. I think dropping explicit declaration in deference to a convention is a better idea for the reasons you have noted. In practice, I wonder if there are likely to be more variables with relative or standard indexing? Please could this behaviour be added to the documentation?